lib/msal-browser/src/cache/LocalStorage.ts

```diff
@@ -37,4 +203,147 @@ export class LocalStorage implements IWindowStorage<string> {
     containsKey(key: string): boolean {
         return window.localStorage.hasOwnProperty(key);
     }
+
+    /**
+     * Removes all known MSAL keys from the cache
+     */
+    clear(): void {
+        // Removes all remaining MSAL cache items
+        this.memoryStorage.clear();
+
+        const accountKeys = getAccountKeys(this);
+        accountKeys.forEach((key) => this.removeItem(key));
+        const tokenKeys = getTokenKeys(this.clientId, this);
+        tokenKeys.idToken.forEach((key) => this.removeItem(key));
+        tokenKeys.accessToken.forEach((key) => this.removeItem(key));
+        tokenKeys.refreshToken.forEach((key) => this.removeItem(key));
+
+        // Clean up anything left
+        this.getKeys().forEach((cacheKey: string) => {
+            if (
+                cacheKey.startsWith(Constants.CACHE_PREFIX) ||
+                cacheKey.indexOf(this.clientId) !== -1
+            ) {
+                this.removeItem(cacheKey);
+            }
+        });
+    }
+
+    /**
+     * Helper to decrypt all known MSAL keys in localStorage and save them to inMemory storage
+     * @returns
+     */
+    private async importExistingCache(correlationId: string): Promise<void> {
+        if (!this.encryptionCookie) {
+            return;
+        }
+
+        const accountKeys = getAccountKeys(this);
+        await this.importArray(accountKeys, correlationId);
+
+        const tokenKeys: TokenKeys = getTokenKeys(this.clientId, this);
+        await Promise.all([
+            this.importArray(tokenKeys.idToken, correlationId),
+            this.importArray(tokenKeys.accessToken, correlationId),
+            this.importArray(tokenKeys.refreshToken, correlationId),
+        ]);
+    }
+
+    /**
+     * Helper to decrypt and save cache entries
+     * @param key
+     * @returns
+     */
+    private async getItemFromEncryptedCache(
+        key: string,
+        correlationId: string
+    ): Promise<string | null> {
+        if (!this.encryptionCookie) {
+            return null;
+        }
+
+        const rawCache = this.getItem(key);
+        if (!rawCache) {
+            return null;
+        }
+
+        let encObj: EncryptedData;
+        try {
+            encObj = JSON.parse(rawCache);
+            if (!encObj.id || !encObj.nonce || !encObj.data) {
+                // Data is not encrypted, likely from old version of MSAL. It must be removed because we don't know how old it is.
+                this.performanceClient.incrementFields(
+                    { unencryptedCacheCount: 1 },
+                    correlationId
+                );
+                throw createBrowserAuthError(
```

Comment:
The throw here and throw on 290 just below can likely be refactored to be just a boolean.  That would likely be faster at runtime than throwing an exception that's just caught right below on 294.  I expect one of these will throw pretty frequently.

Reply:
The try/catch is for the JSON.parse, throwing for the result validations was just a way to consolidate how we handle the conditions to avoid duplication

Reply:
Moved the invalid cache handling into the promise chain for best of both worlds